src/adt.rs

```diff
@@ -140,4 +322,151 @@ pub(crate) mod tests {
+
+            fn batch_read(
+                &self,
+                addresses: Vec<Self::Address>,
+            ) -> impl Send + Future<Output = Result<Vec<Option<Self::Word>>, Self::Error>>
+            {
+                let storage = Arc::clone(&self.storage);
+                async move {
+                    let storage = storage.lock().unwrap();
+                    Ok(addresses
+                        .into_iter()
+                        .map(|addr| storage.get(&addr).cloned())
+                        .collect())
+                }
+            }
+
+            fn guarded_write(
+                &self,
+                guard: (Self::Address, Option<Self::Word>),
+                tasks: Vec<(Self::Address, Self::Word)>,
+            ) -> impl Send + Future<Output = Result<Option<Self::Word>, Self::Error>> {
+                let storage = Arc::clone(&self.storage);
+                async move {
+                    let mut storage = storage.lock().unwrap();
+                    let (guard_addr, guard_value) = guard;
+                    let current_value = storage.get(&guard_addr).cloned();
+                    if current_value == guard_value {
+                        for (addr, word) in tasks {
+                            storage.insert(addr, word);
+                        }
+                    }
+                    Ok(current_value)
+                }
+            }
+        }
+
+        /// Tests the atomicity of the guarded write operation.
+        pub async fn test_guarded_write_atomicity<T>(memory: T)
+        where
+            T: MemoryADT<Address = u8, Word = u8> + Send + Sync + 'static + Clone,
+            T::Error: std::fmt::Debug,
+        {
+            let n: usize = 100; // number of elements to be written in the memory
+            let workers: usize = 25; // number of concurrent workers
+
+            let addresses: Vec<u8> = (0..n).map(|i| i as u8).collect::<Vec<_>>();
+
+            let old = block_on(memory.batch_read(addresses.clone())).unwrap();
+
+            let handles = (0..n)
+                // For the sake of simplicity, we will the address value in its own memory space.
+                .map(|i| (i as u8, i as u8))
+                .collect::<Vec<_>>()
+                // Split the values into chunks of "workers" size
+                .chunks_exact(workers)
+                .map(|vals| {
+                    let vals = vals.to_vec(); // [0->99] && [100->199]
+                    let mem = memory.clone(); // A reference to the same memory
+
+                    // Spawn a new task
+                    tokio::spawn(async move {
+                        for (adr, val) in vals {
+                            // All concurrent tasks will try to write to write to the same address !!!
+                            // As the operation is atomic, only one of them should succeed.
+                            // If two of them succeed, then the implementation is incorrect and guarded write is not atomic.
```

Comment:
To my understanding, your 4 workers will try to write at addresses i*25 to (i+1)*25-1 where i is the worker ID, conditionally to the fact that no value is stored at the default address which should be 0 since you are using `u8` as address type. If the order in which the memory receives the worker's message by decreasing worker ID then all workers should succeed. You will only see one writing if the memory received the message by increasing worker ID... why should be what happens most of the time since you only have 4 workers, the first one to be launched is the ID 0, and you are testing an in-memory Memory.

Try reverse the iterator, see if I'm wrong.